### PR TITLE
[c10d] PT1 Distributed Release MileStone No.1 - Completed Distributed Package and CI tests

### PR DESCRIPTION
**Pull request by teng-li**
The PR includes:
(1) torch.distributed.c10d, which now includes the complete backward compatible frontend API for `torch.distributed`
(2) `env://` init method functionality
(3) Minor change to `test_distributed.py`, which is now a test for `torch.distributed.c10d`.
(4) The old `test_distributed.py' is now moved to `test_distributed_thd`
(5) Miscellaneous bug fixes.
(6) DDP CPU test is removed since c10d doesn't have this support yet, but this is a very easy test after moving DDP CPU's dependency to torch.distributed.c10d.
(7) CI config to test MPI, NCCL, and Gloo backend of c10d

**Now all the distributed test including c10d DDP can pass with the c10d frontend API**

TODO: (in a separate PR)
MPI subgroup support, once this is added, CI group test will be enabled.

### CODE REVIEW

**Review by pietern — APPROVED** [review]
LGTM, aside from minor snafu. There is also the illegal memory access error for the NCCL tests.

File: torch/lib/c10d/ProcessGroupMPI.cpp [hunk 1]
```diff
@@ -449,17 +450,17 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::scatter(
     std::vector<std::vector<at::Tensor>>& inputTensors,
     const ScatterOptions& opts) {
   checkSingleTensor(outputTensors);
+  if (inputTensors.size() != 1) {
+    throw std::runtime_error("Gather: multi-GPU collective is not supported");
```
[the remark this answers is not in the document]
Comment (teng-li) [reply torch/lib/c10d/ProcessGroupMPI.cpp]: fixed

**Comment by pietern** [comment]
Assuming this will turn green when #10932 is merged?

**Comment by teng-li** [comment]
@pietern  One more thing to do, is to fix the c10d ddp test, and hopefully it will be green

**Pull request closed without merging**